postgis/lwgeom_box3d.c

```diff
@@ -206,62 +206,194 @@ Datum BOX3D_to_LWGEOM(PG_FUNCTION_ARGS)
 		pt.y = box->ymin;
+		pt.z = box->zmin;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmin;
+		pt.y = box->ymin;
+		pt.z = box->zmax;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmin;
+		pt.y = box->ymax;
+		pt.z = box->zmax;
 		ptarray_append_point(pa, &pt, LW_TRUE);
 		pt.x = box->xmin;
 		pt.y = box->ymax;
+		pt.z = box->zmin;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmin;
+		pt.y = box->ymin;
+		pt.z = box->zmin;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+
+		ppa[0] = pa;
+		lwpoly = lwpoly_construct(SRID_UNKNOWN, NULL, 1, ppa);
+		result = geometry_serialize(lwpoly_as_lwgeom(lwpoly));
+		lwpoly_free(lwpoly);
+	}
+	else if (box->ymin == box->ymax)
+	{
+		LWPOLY *lwpoly;
+		POINTARRAY **ppa = lwalloc(sizeof(POINTARRAY*));
+
+		pt.x = box->xmin;
+		pt.y = box->ymin;
+		pt.z = box->zmin;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmin;
+		pt.y = box->ymin;
+		pt.z = box->zmax;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmax;
+		pt.y = box->ymin;
+		pt.z = box->zmax;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmax;
+		pt.y = box->ymin;
+		pt.z = box->zmin;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmin;
+		pt.y = box->ymin;
+		pt.z = box->zmin;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+
+		ppa[0] = pa;
+		lwpoly = lwpoly_construct(SRID_UNKNOWN, NULL, 1, ppa);
+		result = geometry_serialize(lwpoly_as_lwgeom(lwpoly));
+		lwpoly_free(lwpoly);
+	}
+	else if (box->zmin == box->zmax)
+	{
+		LWPOLY *lwpoly;
+		POINTARRAY **ppa = lwalloc(sizeof(POINTARRAY *));
+
+		pt.x = box->xmin;
+		pt.y = box->ymin;
+		pt.z = box->zmin;
+		ptarray_append_point(pa, &pt, LW_TRUE);
+		pt.x = box->xmin;
+		pt.y = box->ymax;
+		pt.z = box->zmin;
 		ptarray_append_point(pa, &pt, LW_TRUE);
 		pt.x = box->xmax;
 		pt.y = box->ymax;
+		pt.z = box->zmin;
 		ptarray_append_point(pa, &pt, LW_TRUE);
 		pt.x = box->xmax;
 		pt.y = box->ymin;
+		pt.z = box->zmin;
 		ptarray_append_point(pa, &pt, LW_TRUE);
 		pt.x = box->xmin;
 		pt.y = box->ymin;
+		pt.z = box->zmin;
 		ptarray_append_point(pa, &pt, LW_TRUE);
 
+		ppa[0] = pa;
+		lwpoly = lwpoly_construct(SRID_UNKNOWN, NULL, 1, ppa);
 		result = geometry_serialize(lwpoly_as_lwgeom(lwpoly));
-		
+		lwpoly_free(lwpoly);
+	}
+	else
+	{
+		POINT4D points[8];
+		LWGEOM **geoms = NULL;
+		LWGEOM *geom = NULL;
+		int ngeoms = 6;
+
+		/* Initialize the 8 vertices of the box */
+		points[0] = (POINT4D) { box->xmin, box->ymin, box->zmin };
+		points[1] = (POINT4D) { box->xmin, box->ymax, box->zmin };
+		points[2] = (POINT4D) { box->xmax, box->ymax, box->zmin };
+		points[3] = (POINT4D) { box->xmax, box->ymin, box->zmin };
+		points[4] = (POINT4D) { box->xmin, box->ymin, box->zmax };
+		points[5] = (POINT4D) { box->xmin, box->ymax, box->zmax };
+		points[6] = (POINT4D) { box->xmax, box->ymax, box->zmax };
+		points[7] = (POINT4D) { box->xmax, box->ymin, box->zmax };
+
+		geoms = (LWGEOM **) lwalloc(sizeof(LWGEOM *) * ngeoms);
```

Comment:
Agreed -- too much knowledge of the `LWCOLLECTION` data type.  Let's leave as-is.